apps/labs/posts/unlocking-c-level-performance-in-df-apply.md

```diff
@@ -0,0 +1,466 @@
+---
+title: 'Unlocking C-level performance in pandas DataFrame.apply with numba'
+authors: [thomas-li]
+published: October 15, 2023
+description: 'A high-level overview of the new numba engine in DataFrame.apply'
+category: [PyData ecosystem]
+featuredImage:
+  src: /posts/numpy-python-api-cleanup/numpy-python-api-cleanup-featured.png
+  alt: 'todo'
+hero:
+  imageSrc: /posts/numpy-python-api-cleanup/numpy-python-api-cleanup-hero.png
+  imageAlt: 'todo'
+---
+
+## Introduction
+
+If you've been using pandas for a while, you are probably familiar with the family of `apply` functions in pandas (e.g. `DataFrame.apply`, `Series.apply`, `DataFrameGroupBy.apply`, etc.). For those unaware, these functions allow you to specify a Python function to be called for each row/column of a DataFrame, and then combines and returns the results back to you.
+
+Because everything happens in Python-land, though, apply is usually very slow, and its recommended that you only use it when you cannot find an equivalent
+method that does what you want.
+
+In pandas 2.2.0, though, a new engine (`engine="numba"`) option will be added to `DataFrame.apply`, opening up the possibility for fast and parallelizable apply in pandas. It works by using numba, a JIT (just-in-time) compiler that translates your Python/numpy functions into fast machine code when their called, providing a several times speedup compared to the original python engine.
+
+So, can I just slap on `engine="numba"` to `DataFrame.apply` and enjoy a free many-times speedup?
+
+```diff
+- df.apply(f)
++ df.apply(f, engine="numba")
+```
+
+Well, not quite. Because of the way that numba/the numba engine inside pandas works, there are several caveats and gotchas to keep in mind to obtain this speedup.
+
+TLDR:
+Use apply only when you have to (e.g. there is no corresponding method or chain of pandas methods that do that you want).
+
+## How the numba engine works
+
+Currently, in pandas, there are currently two implementations of the numba engine inside of apply, one for `raw=True`, and one for `raw=False` (the default).
+
+If you didn't know already, passing in `raw=True` to `DataFrame.apply` tells pandas to pass in numpy arrays, not pandas Series to the functions you pass into apply. Because, numba itself supports numpy arrays natively, in general, every function runnable with the `Python` engine, should also be runnable with the `numba` engine, as long as you stick to the [supported numpy features](https://numba.readthedocs.io/en/latest/reference/numpysupported.html) inside numba.
+
+When `raw=False` is passed in or no `raw` parameter is passed in, we must "mock" the pandas `Series` object that would normally get passsed in, since numba doesn't natively support pandas DataFrames. This means that while most functions passed into apply with work with the numba engine out of the box with `raw=False`, some functions may require modifications, as the "mocked" DataFrame does not implement the entire pandas API.
```

Comment:
I think a lot of the complexity of the work came from this mocked Series/DataFrame implementation, right? It would be nice to say a bit more about it. You did some quite interesting work there I believe, however this doesn't really come out in this post yet. So 1-2 paragraphs about some of the key things you had to implement or work around in order to get things to work would help the story I think. Plus pointing at one or a couple of issues/PRs can't hurt.

Reply:
I added a link to the implementation.

There really aren't a lot of issues/PRs taling about this, I basically implemented this in one go in a PR.
(Not sure if there's value linking to my failed attempts here for Groupby.apply. I basically got halfway through on those before I got stuck and had to stop.)